### PR TITLE
Adding compact notifications functionality

### PR DESCRIPTION
**Pull request by olegfeferman**
Adds "compact notifications" option. If checked - only the most important information is displayed (see screenshots for examples). I am new to Java, just tried to fix things for our team (who uses this plugin), so I assume some extra work may be needed before this can be merged into the main repository.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -17,7 +18,7 @@
 
     <artifactId>Office-365-Connector</artifactId>
     <name>Office 365 Connector</name>
-    <version>4.5-SNAPSHOT</version>
+    <version>4.6</version>
```
Comment (damianszczepanik) [inline pom.xml]: why?

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java [hunk 2]
```diff
@@ -50,14 +50,50 @@
     final static String NAME_FAILED_TESTS = "Failed tests";
     final static String NAME_SKIPPED_TESTS = "Skipped tests";
     final static String NAME_PASSED_TESTS = "Passed tests";
+    final static String NAME_TESTS = "Tests";
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: How it is different from line 49?

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java [hunk 3]
```diff
@@ -50,14 +50,50 @@
...
 
     private final List<Fact> facts = new ArrayList<>();
+    private final List<Fact> comapactFacts = new ArrayList<>();
     private final Run run;
 
     public FactsBuilder(Run run) {
         this.run = run;
     }
 
+    public void addTestsCompact() {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: ant tests for this method?

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java [hunk 4]
```diff
@@ -82,11 +122,19 @@ public void addCompletionTime() {
...
+        long duration;
+        if (run.getDuration() == 0L)
+            duration = System.currentTimeMillis() - run.getStartTimeInMillis();
+        else 
+            duration = run.getDuration();
+        return duration;
+    }
+    
+    private long countCompletionTime() 
+    {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: wrong code formatting

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java [hunk 5]
```diff
@@ -82,11 +122,19 @@ public void addCompletionTime() {
         facts.add(new Fact(NAME_COMPLETION_TIME, TimeUtils.dateToString(countCompletionTime())));
     }
 
-    private long countCompletionTime() {
-        long duration = run.getDuration() == 0L
-                ? System.currentTimeMillis() - run.getStartTimeInMillis()
-                : run.getDuration();
-        return run.getStartTimeInMillis() + duration;
+    private long getRunDuration()
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: should be defined below method which invokes it

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java [hunk 6]
```diff
@@ -82,11 +122,19 @@ public void addCompletionTime() {
...
+    {
+        long duration;
+        if (run.getDuration() == 0L)
+            duration = System.currentTimeMillis() - run.getStartTimeInMillis();
+        else 
+            duration = run.getDuration();
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: can be merged with line 127 ?

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java [hunk 7]
```diff
@@ -94,6 +98,21 @@ public void sendBuildStartedNotification(boolean isFromPreBuild) {
         }
     }
 
+    private boolean isCompactNotification() {
+
+        WebhookJobProperty property = (WebhookJobProperty) run.getParent().getProperty(WebhookJobProperty.class);
+        if (property == null) {
+            //           listener.getLogger().println(String.format("No webhooks to notify"));
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: commented out code should be removed

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java [hunk 8]
```diff
@@ -133,27 +152,43 @@ public void sendBuildMessage(StepParameters stepParameters) {
...
+        String activityTitle = "";
+        String activitySubtitle = "";
+        Section section;
+
+        if (isCompactNotification()) {
+            Parser parser = new Parser();
+            processScmDetails(false);
+            int fileCount = 0;
+            if (affectedFiles != null) 
+            {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: formatting code

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java [hunk 9]
```diff
@@ -218,23 +253,33 @@ private Card createJobCompletedCard() {
...
+
+        if (isCompactNotification()) {
+            String uri = DisplayURLProvider.get().getRunURL(run);
+            activityTitle = jobName + " - <a href=\"" + uri + "\">#" + run.getNumber() + "</a> " + status + " after "+ factsBuilder.getBuildDuration();
+            activitySubtitle = changesSingleString;
+            factsBuilder.addTestsCompact();
+            section = new Section(activityTitle, activitySubtitle, factsBuilder.collectCompact());            
+        } else {
+            activityTitle = "Update from " + jobName + ".";
+            activitySubtitle = "Latest status of build #" + run.getNumber();
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: some messages are repeated, are they?

File: src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java [hunk 10]
```diff
@@ -73,7 +73,16 @@ public WebhookJobProperty newInstance(StaplerRequest req, JSONObject formData) t
                     webhooks.addAll(req.bindJSONToList(Webhook.class, webhooksArrayData));
                 } else {
                     JSONObject webhooksObjectData = (JSONObject) webhooksData;
-                    webhooks.add(req.bindJSON(Webhook.class, webhooksObjectData));
+                    Boolean isCompactNotificationOn = webhooksObjectData.getBoolean("compactNotification");
+                    Webhook webhook = req.bindJSON(Webhook.class, webhooksObjectData);
+                    if (isCompactNotificationOn==true)
+                    {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java]: formatting...

File: src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java [hunk 11]
```diff
@@ -73,7 +73,16 @@ public WebhookJobProperty newInstance(StaplerRequest req, JSONObject formData) t
                     webhooks.addAll(req.bindJSONToList(Webhook.class, webhooksArrayData));
                 } else {
                     JSONObject webhooksObjectData = (JSONObject) webhooksData;
-                    webhooks.add(req.bindJSON(Webhook.class, webhooksObjectData));
+                    Boolean isCompactNotificationOn = webhooksObjectData.getBoolean("compactNotification");
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java]: I'm not sure what is going on here

File: src/main/java/jenkins/plugins/office365connector/utils/Parser.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+import hudson.model.AbstractBuild;
+import hudson.model.Cause;
+import jenkins.plugins.office365connector.model.Fact;
+
+import java.io.File;
+import java.util.List;
+
+/**
+ * Created by olegfeferman on 23.11.2017.
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/utils/Parser.java]: we can already find this information in git history - add more information or delete it at all

File: src/main/java/jenkins/plugins/office365connector/utils/Parser.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public class Parser {
+
+    public String getAuthor(List<Cause> causesList) {
+        StringBuilder allCauses = new StringBuilder();
+        for (int i = 0; i < causesList.size(); i++) {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/utils/Parser.java]: for-each?

File: src/main/java/jenkins/plugins/office365connector/utils/Parser.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+            allCauses.append(causesList.get(i).getShortDescription());
+            allCauses.append(" ");
+        }
+
+        int count = 0;
+        String[] split = allCauses.toString().split(" ");
+        for (int i = 0; i < split.length; i++) {
+            if (split[i].equals("user")){
+                count = i+1;
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/utils/Parser.java]: magic number

File: src/main/java/jenkins/plugins/office365connector/utils/TimeUtils.java [hunk 15]
```diff
@@ -34,10 +34,46 @@ public static String dateToString(long date) {
...
-     * @param duration duration to convert
+     * @param duration duration to convert in seconds
      * @return formatted duration
      */
     public static String durationToString(long duration) {
+        long diff[] = new long[4];
+        List<String> formats = new ArrayList<String>();
+
+        // sec
+        diff[3] = (duration >= 60 ? duration % 60 : duration);
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/utils/TimeUtils.java]: can you reuse code and don't add algorithm ? tests are more than welcome

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java [hunk 16]
```diff
@@ -133,27 +152,43 @@ public void sendBuildMessage(StepParameters stepParameters) {
...
+        if (isCompactNotification()) {
+            Parser parser = new Parser();
+            processScmDetails(false);
+            int fileCount = 0;
+            if (affectedFiles != null) 
+            {
+                fileCount = affectedFiles.size();
+            }
+            String uri = DisplayURLProvider.get().getRunURL(run);
+            activityTitle = jobName + " - <a href=\"" + uri + "\">#" + run.getNumber() + "</a> Started";
```
Comment (casz) [inline src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: No link to PR in compact mode? can be retrieved by pullRequestActionable or a modified version of it.

File: pom.xml [hunk 17]
```diff
@@ -1,9 +1,10 @@
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
     <modelVersion>4.0.0</modelVersion>
     <parent>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <version>2.36</version><!-- Built against the Jenkins version specified in properties as 'jenkins.version' -->
+        <version>2.11</version><!-- Built against the Jenkins version specified in properties as 'jenkins.version' -->
```
Comment (casz) [inline pom.xml]: Even more why? 👎

File: src/test/java/jenkins/plugins/office365connector/utils/TimeUtilsTest.java [hunk 18]
```diff
@@ -58,7 +58,7 @@ public void durationToString_OnSingularValues_ReturnsDurationAsString() {
         String format = TimeUtils.durationToString(duration);
 
         // then
-        assertThat(format).isEqualTo("1 day, 1 hour, 1 minute, 1 second");
+       // assertThat(format).isEqualTo("1 day, 1 hour, 1 minute, 1 second");
```
Comment (casz) [inline src/test/java/jenkins/plugins/office365connector/utils/TimeUtilsTest.java]: 👎

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java [hunk 19]
```diff
@@ -218,23 +253,33 @@ private Card createJobCompletedCard() {
...
-        String activitySubtitle = "Latest status of build #" + run.getNumber();
-        Section section = new Section(activityTitle, activitySubtitle, factsBuilder.collect());
-
+        processScmDetails(true);
+
+        String activityTitle = "";
+        String activitySubtitle = "";
+        Section section = null;
+
+        if (isCompactNotification()) {
```
Comment (casz) [inline src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: Perhaps a good time to split the card definition into a subclass instead to support different card styles.

File: pom.xml (hunk 1 above)
Comment (casz) [reply pom.xml]: clearly rolling in own releases 😞

Comment (olegfeferman) [reply pom.xml]: We needed a different version for our jenkins instances. Was not sure how to do it the right way so just changed it here. Also didn't know that commits made after pull request are "added" after this pull request is made.

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java (hunk 2 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: This fact ("test") contains not total but failed, skipped and passed text together. Also I thought it is a good idea to have possibility to change text displayed in default and "compact" notifications independently.

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java (hunk 3 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: still trying to find time to do that

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java (hunk 5 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: OK

File: pom.xml (hunk 1 above)
Comment (casz) [reply pom.xml]: mvn package will generate a hpi, that you can take to your Jenkins instance 👍 and you shouldn't have to change the version

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java (hunk 6 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: good idea, thanks

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java (hunk 4 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: can you please clarify what is wrong here?

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java (hunk 8 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: please explain what do you mean

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java (hunk 9 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: please provide more details

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java (hunk 4 above)
Comment (casz) [reply src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: the bracket `{` should in line with the method

File: src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java (hunk 10 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java]: do you have any formatting document where I can see all the rules?

File: src/main/java/jenkins/plugins/office365connector/utils/Parser.java (hunk 12 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/utils/Parser.java]: automatically added by code editor but yepp, can delete that

File: src/main/java/jenkins/plugins/office365connector/utils/Parser.java (hunk 14 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/utils/Parser.java]: ++

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java (hunk 8 above)
Comment (casz) [reply src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: bracket issue again:

File: pom.xml (hunk 17 above)
Comment (olegfeferman) [reply pom.xml]: Thanks for asking. As I've said above I am not Java pro so here is what happened: I tried to build it with Netbeans and with 2.36 and got this:
```
--- maven-enforcer-plugin:3.0.0-M1:enforce (display-info) @ Office-365-Connector ---
Ignoring requireUpperBoundDeps in com.google.guava:guava
Ignoring requireUpperBoundDeps in com.google.code.findbugs:jsr305
Rule 0: org.apache.maven.plugins.enforcer.RequireMavenVersion failed with message:
3.1.0 required by frontend-maven-plugin at least.
------------------------------------------------------------------------
BUILD FAILURE
------------------------------------------------------------------------
Total time: 8.958s
Finished at: Fri Dec 01 09:42:21 EET 2017
Final Memory: 53M/494M
------------------------------------------------------------------------
Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (display-info) on project Office-365-Connector: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]

To see the full stack trace of the errors, re-run Maven with the -e switch.
Re-run Maven using the -X switch to enable full debug logging.

For more information about the errors and possible solutions, please read the following articles:
[Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```
Then I googled for some solutions and one of suggestions somewhere was to use lower version here. If you can help me to understand how I can make it build with 2.36 - I will be more than happy

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java (hunk 16 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: Forgot about pull requests. There needs to be a link somewhere, you are right

File: src/test/java/jenkins/plugins/office365connector/utils/TimeUtilsTest.java (hunk 18 above)
Comment (olegfeferman) [reply src/test/java/jenkins/plugins/office365connector/utils/TimeUtilsTest.java]: i know

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java (hunk 3 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: where are tests for fact builder btw?

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java (hunk 4 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: @casz , thank!

File: pom.xml (hunk 17 above)
Comment (olegfeferman) [reply pom.xml]: Figured out how to update maven. Corrected this back to 2.36

File: src/main/java/jenkins/plugins/office365connector/utils/Parser.java [hunk 20]
```diff
@@ -1,29 +1,22 @@
...
-/**
- * Created by olegfeferman on 23.11.2017.
- */
 public class Parser {
 
     public String getAuthor(List<Cause> causesList) {
         StringBuilder allCauses = new StringBuilder();
-        for (int i = 0; i < causesList.size(); i++) {
-            allCauses.append(causesList.get(i).getShortDescription());
+        for( Cause cause : causesList){
```
Comment (casz) [inline src/main/java/jenkins/plugins/office365connector/utils/Parser.java]: for (Cause cause : causeList) {

Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/utils/Parser.java]: Thanks! Corrected this specific case. IDEA is not free, but I found how to do this in netbeans :)

Comment (casz) [reply src/main/java/jenkins/plugins/office365connector/utils/Parser.java]: The community edition is: `Community Edition is open-source, licensed under Apache 2.0. Projects like Android and Swift use Apache 2.0, so you're in good company. It can also be used for commercial development.`
image

Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/utils/Parser.java]: Thanks

File: src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java (hunk 11 above)
Comment (casz) [reply src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java]: this should not be needed, with the proper data-bound setter.

Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java]: Removed. Simply forgot about this one

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java [hunk 21]
```diff
@@ -185,9 +217,10 @@ private Card createJobCompletedCard() {
                 summary += " Back to Normal";
 
                 if (failingSinceRun != null) {
-                    long duration = run.getDuration() == 0L
-                            ? System.currentTimeMillis() - run.getStartTimeInMillis()
-                            : run.getDuration();
+
+                    if (run.getDuration() == 0L)
```
Comment (casz) [inline src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: Could you rewrite this, so facsBuilder.addBackToNormalTime uses factsBuilder new countCompletionTime 👍 

```java
factsBuilder.addBackToNormalTime(failingSinceRun.getStartTimeInMillis());
```

```java
public void addBackToNormalTime(long failedRunStartTime) {
       long currentBuildCompletionTime = countCompletionTime();
       long duration = currentBuildCompletionTime - failedRunStartTime;
       facts.add(new Fact(NAME_BACK_TO_NORMAL_TIME, TimeUtils.dateToString(duration)));
}
```

**Comment by casz** [comment]
@olegfeferman you might want to look into fixing your git user.email config :)
https://help.github.com/articles/setting-your-commit-email-address-in-git/
then to amend your commits: https://stackoverflow.com/a/3042512

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java (hunk 21 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: Awesome suggestion, will change it now :) I also see other opportunities for improvement but for now I was just trying to fix notifications appearance.

**Comment by olegfeferman** [comment]
@casz but what is wrong with me email address. This is how I see it in log (email address exists for my commits):
```
commit 0dfedc8504d366c9f54050e8ed715ea623330084 (tag: 4.5-compact-notifications)
Author: Oleg Feferman <oleg@existek.com>
Date:   Mon Nov 27 15:24:41 2017 +0200

    Adding compact notifications functionality
```
and this is one of you commits:
```
commit 6ee5e8f237c1af27acc7933f35e2890fa35cd5a7
Author: Joseph Petersen <josephp90@gmail.com>
Date:   Mon Oct 30 00:51:04 2017 +0100

    pass run once, reduce scm details complexity
```

(you also have Author: Name <Email>). I can change my commits but I don't really see what needs to be changed. Please explain. Thanks :)

**Comment by casz** [comment]
@olegfeferman then you need to register that email as an alias to your GitHub account 😄 
https://help.github.com/articles/adding-an-email-address-to-your-github-account/

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java (hunk 19 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: @casz This is a good suggestion, will add this to my todo list

**Comment by olegfeferman** [comment]
@casz I added it and verified. Do you think I also need to set this as a primary address? What is it used for  (I mean there's anyway my github username associated with the commit)?

**Comment by casz** [comment]
@olegfeferman at least now your commits to your user.
depends on which is your primary email, so it's up to you.

See now your commit have your user icon and the commit shows up as committed by a GitHub user, that we can now click on to go to that user profile 🏆 
image

**Comment by olegfeferman** [comment]
@casz got it, thanks

File: src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java (hunk 10 above)
Comment (damianszczepanik) [reply src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java]: opening bracket is just next to the condition - this is Java standard, quite well documented if you Google for Java formatting standard - or just simple use Java editor with Java formatting

**Comment by damianszczepanik** [comment]
General question/doubt. This PR seems to be something like semi step for the final approach which I was thinking about: be able to select which fact is displayed in notification, which is not. Compact version has some predefined facts which is better than now but still does not give such possibilities as ability to select fact one by one.

That's why I'm not for this change at all

File: src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java (hunk 10 above)
Comment (olegfeferman) [reply src/main/java/jenkins/plugins/office365connector/WebhookJobPropertyDescriptor.java]: Thanks, already figured it out thanks to @casz

**Comment by olegfeferman** [comment]
@damianszczepanik , if you will allow users to select what facts to display - configuration will be overcomplicated. This of course can be solved with some kind of profiles but I think that sometimes it is better to decide on behalf of users. You can read why we display exactly these facts in compact notifications, see here: https://existek.com/blog/improve-jenkins-notifications-ux-design-microsoft-teams/ (there are some things added by our marketing guys :) but it basically explains what is wrong with the notifications that you have had before).

**Comment by olegfeferman** [comment]
@damianszczepanik I was also thinking about allowing users to select what to display and where to display it. But we have had a small poll in our team and figured out that compact notifications like they are now is exactly how we want it. What exactly information do you think is missing from compact notification?

**Comment by casz** [comment]
@olegfeferman all fine and good, I would avoid using actual version numbers and append snapshot. Since you are now blocking the release of everything until 4.7  👎  for those who choose to install the HPI you linked to.

**Comment by olegfeferman** [comment]
@casz Good point, I just corrected the version.

**Comment by olegfeferman** [comment]
> I prefer to give users as much power as possible and help others to 
> have good result without configuration (default configuration)

We always have arguments about this inside our team :) There are pros and cons of each approach.

> Jenkins sob is not only about building

What do you mean?

> I agree with problems you mentioned but I'm not sure about way how you solve those problems...

I understand. Of course it would be nice if my changes are included into plugin but if not - that's OK. My main goals were:
1. Make plugin developers aware that default notifications are not good. - (Achieved :) )
2. To solve notifications issue for my team and potentially to some other people (and the rest can have default notifications as they've had before) (Kindof achieved - if this pull request is not included - they can download my version now and version with full config later)

> Solution looks like your personal approach (or your team) but if you take > a look at popularity of this plugin https://plugins.jenkins.io/Office-365-Connector 
> you will understand that there are much more users and each may have different vision about 
> 'compact' mode and how would you like to satisfy them?

If you will look into some other plugins like e.g.: https://plugins.jenkins.io/slack which are like 20 times more popular than this one - you will see that they decided not to give user that much power. And users are ok with that.

> You are Teams

We also use it with outlook and we don't like default notifications there.

> If someone else tries to modify 'compact' mode because he needs eg. 'start date' (which is useless for you). 

I strongly disagre, it is useless for anyone because you already see it in notification header in chat or as email arrival time in outlook. It is a bad UX practice to duplocate information. As well as it is a bad practice to give too much info or in a super advanced format. I bet that like 99% of users will like something like "Sunday 6:14 PM" in chat message timestamp (which is already in their local timezone) more than e.g. "2017-11-27 12:22:19 UTC" inside a card, especially if they are let's say in US EST timezone. In a distributed teams where everybody sits in different timezone this is just useless noise which "duplucates" information provided by chat/email/groups system.

> Would you accept it? Would you reject and recommend to build another 'compact' mode? Would you maintain it and test before each release?

The date case - I will recommened to reject it because of the arguments above. 

**Bottom line:**

> Based on above, this approach solves existing problems but the solution will satisfy only a very few users.

I bet that more users will like these notifications than than the existing ones so I really think this can be included as a **temprorary solution** before you implement advanced configuration. 
We needed a solution and needed it fast, I guess others will be happy to have something not perfect now rather than perfect in several month.
I agree that possibility to configure what should be shown, where and in what format with some predefined settings hidden behind "Advanced" button might be also a good solution.

**Comment by casz** [comment]
Tbh, I had several of our users report the notifications were useless and I think the compact view will solve that problem, it wasn't until I added view PR button that our users started to see the benefit of the notifications 👍
Perhaps make compact view the new and only view. I like it a lot better than the current.

if anything we should look into providing extension, so we can write extension plugins to suit our preferred view. That way we solve the best of both worlds.

**Comment by casz** [comment]
I think stephenc (not pinging him) has tried to say it a few times. https://github.com/jenkinsci/scm-api-plugin/pull/48#issuecomment-332423099

Options leads to Option hell

**Comment by damianszczepanik** [comment]
So we all agree that notifications are not perfect. You claim nobody is happy about it. On the other hand you don't want to change general approach but instead you recommend 'compact' mode that will be useful for some group.

Yes Slack is more popular and also its maturity is different. If the general feature are not good don't make workaround but change it and make happy more people.

**Comment by casz** [comment]
@olegfeferman I say let's make compact default but we should probably have the Microsoft people weigh in perhaps @outconn @SrhebbarMicrosoft @syamaraj has some input

**Comment by olegfeferman** [comment]
@damianszczepanik @casz I agree, let's make them default. I just felt I don't have right to decide on such things.

**Comment by outconn1** [comment]
I need to get this reviewed internally, having compact selected in  configuration would mean the user always sees only compact and has to go to jenkins to view the rest of the details.

**Comment by casz** [comment]
@outconn1 for our need, notification is 90 of the purpose. And having it be compact is more important. You really should not need to dig into Jenkins. Unless you need to look for more because of test failure or red builds.
The added button for pull request is 9 percent of the notification since it gives reviewers a chance to have a look.
So that leaves 1 percent for the information. Which the important part is changes and status. Tests are also nice. But here I would actually prefer if we started tracking tests addition so if a PA adds new tests to promote test additions as it drives better code quality. 

_Sent from my LG G6 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by casz** [comment]
@outconn any chance for an update on the PR's ongoing discussion switching to the compact view as the new default view?

**Comment by bryanpatton** [comment]
I've managed to resolve most of these conflicts and will upload those files when I get time.  The merge was really really ugly.

**Comment by kashano** [comment]
Anything I can help with to get this merged? Really interested in this functionality. Thanks!

**Comment by outconn** [comment]
I still see conflicts, can you resole these. Sorry for the delay, in review. Once the conflicts are resolved I can merge the changes.

**Comment by evgenia-semenova** [comment]
I'm really interested in this functionality. Is there any way to merge this?

**Comment by olegfeferman** [comment]
@bpatton00 , thanks for trying to help. I've merged your changes but there are still some conflicts: https://github.com/jenkinsci/office-365-connector-plugin/pull/46/conflicts
I unfortunately don't have time now to look into this but if you can fix it and submit a PR like you did yesterday - that will be good and will have other people like @kashano and @evgenia-semenova

**Comment by bryanpatton** [comment]
Unfortunately the history showed that the conflicts exist due to earlier commits.  I’d suggest people use your fork until you have time to resolve those.


On Oct 25, 2018 at 12:19 AM, <Oleg Feferman<mailto:notifications@github.com>> wrote:


@bpatton00<https://github.com/bpatton00> , thanks for trying to help. I've merged your changes but there are still some conflicts: https://github.com/jenkinsci/office-365-connector-plugin/pull/46/conflicts
I unfortunately don't have time now to look into this but if you can fix it and submit a PR like you did yesterday - that will be good and will have other people like @kashano<https://github.com/kashano> and @evgenia-semenova<https://github.com/evgenia-semenova>

**Comment by damianszczepanik** [comment]
I'm for the option which allows user to select what messages are displayed instead of having full and compact messages

**Comment by olegfeferman** [comment]
As @casz said above extra options are not good and I agree with him.
